### ISSUE TITLE
Update whole store with one JS call

### ISSUE THREAD

**Issue by scorsi** [ryangjchandler/spruce#30]
I'm using dynamically injected html (with Phoenix Live View) and to make AlpineJS works I have to trick some differents things.
Actually I have a injected html node with : 
```html 
<div class="hidden" phx-hook="SpruceStoresHook" phx-data="<%= Jason.encode!(@messages) %>"></div>
```
Which allow me to put server messages variable into the front. I would like to make something like:
```js
Spruce.store('chat_messages') = JSON.parse(this.el.getAttribute('phx-data'));
```
Which is not working. Maybe it would be nice to allow that ? :)

**Comment by HugoDF** [comment]
Have you tried reinitialising the store?

```js
Spruce.store('name', newState)
```

**Comment by scorsi** [comment]
@HugoDF tried but it's not working. You can take a look at the source code to understand why : https://github.com/ryangjchandler/spruce/blob/99255047ce677312024ec674f8f2b8f892bc3ff8/src/index.js#L45.
If the store already existes, it returns it without any modification.

**Comment by HugoDF** [comment]
Yeah, but if that's the case I'm not sure how you're intending to get `store() = newState` to work 😁

Essentially we might need a `store().$update(newState)` function

**Comment by scorsi** [comment]
Actually I achieve that by doing :
```js
const actualizeSpruceComponent = (store_name, new_state) => {
    const store = Spruce.store(store_name);
    for (const key in new_state) {
        if (new_state.hasOwnProperty(key) && store.hasOwnProperty(key)) {
            store[key] = new_state[key];
        }
    }
}
```
Which can be quite expensive on object with multiple keys...

**Comment by ryangjchandler** [comment]
Hi @scorsi, the original thinking behind not letting you overwrite stores is so that it doesn't happen accidentally, since Spruce isn't an immutable solution.

`Spruce.store() = {}` wouldn't work since the function returns a reference to the object, not an address pointer. 

I'd be happy to implement a new function for overwriting the store completely, so that it's clear that's what you're doing, maybe a `Spruce.update()`, or as @HugoDF suggested, a magic `$update` function that is attached to the store itself.

**Comment by HugoDF** [comment]
Bit of a can of worms this one because with a name like `update` people might expect that you're merging the update object into the store (ie. not modifying parts of the store not in the update object)

Maybe `reset` is a better name (and shows that you're wiping away all the state)

**Comment by scorsi** [comment]
I think merging new state to the old state is what we can expect from a state manager. It's likely how other state manager works.
But if we call like `Spruce.store('my_store') = my_new_state` (with the magic `$update` function ?) it could results in misunderstanding so resetting is a better idea.
So people wanting to merge state have to do : `Spruce.store('my_store') = { ...Spruce.store('my_store'), my_new_state }`. But each key not updated will call an update event to other Alpine components ? It have to be optimized.

**Comment by ryangjchandler** [comment]
Yeah, I was planning on just merging the objects together, should be as simple as spreading the two out.

**Comment by scorsi** [comment]
I think `Spruce.store(key, state)` is for initialization only. `Spruce.store('my_store') = my_new_state` with merging option is a better idea in my point of view, a good explanation in the documentation about the "magic merge" have to be explained if you would like to go into that direction.
Maybe `Spruce.store('my_store') = my_state` can also call `Spruce.store('my_store', my_state)` if the store doesn't exists.

**Comment by HugoDF** [comment]
I don't think you'll ever be able to make this work in JS
```
Spruce.store() = state
```

You might be able to get something similar to the following working though:

```
Spruce.storeName = state
```

**Comment by scorsi** [comment]
@HugoDF You're right ^^

**Comment by ryangjchandler** [comment]
Yeah, `Spruce.store() = state` is impossible with JavaScript since pointers are non existent.

**Comment by ryangjchandler** [comment]
I'd much prefer having a method that does this, like `reset()` or `overwrite()`.

**Comment by HugoDF** [comment]
> Yeah, I was planning on just merging the objects together, should be as simple as spreading the two out.
> 
> Or, simply removing the "store is set" check so that you can call `Spruce.store()` as many time as you like, overwriting the store each time.

Yeah this sounds like the lowest effort but it might be compromising on your API.

**Comment by HugoDF** [comment]
> I'd much prefer having a method that does this, like `reset()` or `overwrite()`.

Spruce plugin? 😂😂😂 Seems to be doable in userland in any case

**Comment by scorsi** [comment]
Making `Spruce.storeName = state` could result in big refacto since the `Spruce` object itself should be a `Proxy`.

> I'd much prefer having a method that does this, like `reset()` or `overwrite()`.

is the best solution.

**Comment by ryangjchandler** [comment]
> Making `Spruce.storeName = state` could result in big refacto since the `Spruce` object itself should be a `Proxy`.
> 
> > I'd much prefer having a method that does this, like `reset()` or `overwrite()`.
> 
> This is the best solution. `Spruce.store('my_store').reset(my_new_store)` and `Spruce.store('my_store').overwrite(my_store_to_be_merged)`.

No, I think it should be Spruce.reset('storeName', { ... }), otherwise I'll need to proxy the method calls on each store or attach a method to each store which could conflict with methods that already exist on there.

**Comment by scorsi** [comment]
Since `Spruce.set('storeName', 'key', 'value')` doesn't exists, it's quite weird here.
And in addition, what if I want to do `$spruce.my_store.overwrite({ messages: [/* some heavy object maybe inserted by dynamically injected html */] })` ? :)

**Comment by HugoDF** [comment]
Call it `Spruce.setState` 👀👀👀 hahah confuse everyone with the mix of Vue and React terminology

**Comment by ryangjchandler** [comment]
Realistically, the right thing to do would be not overwriting the entire state object for your store, but instead updating a sub-property of the state instead.

I'm still happy to add an `update()`, `reset()` or `setState()` method.

**Comment by scorsi** [comment]
> Realistically, the right thing to do would be not overwriting the entire state object for your store, but instead updating a sub-property of the state instead.

It was just an example. If you prefer : what if I want to update two keys (or even more) in the same event ? Actually : `$store.my_store.my_first_key = my_first_value; $store.my_store.my_second_key = my_second_value; ......` Using `$refs` in addition could result in a mess.

> I'm still happy to add an update(), reset() or setState() method.

Could be awesome !

**Comment by ryangjchandler** [comment]
Yeah, even if a method it's still going to be nearly as long since you'll need to pass in an object, etc.

Also, if your expressions are really that long I'd suggest moving all of the logic into a separate function.
